lib/read-stream.js

```diff
@@ -167,99 +165,156 @@ ReadStream.prototype = {
 
-      this.blocksRead++
+      this.blocksRead += bytesRead / this.blockSize
       this.bytesRead += bytesRead
       this.position += bytesRead
 
+      debug( 'read-block:blocksRead', this.blocksRead )
+
       // Feed the hash if we're verifying
-      if( this.options.verify ) {
+      if( this.verify ) {
         this._hash.update( buffer )
       }
 
-      // Check if we should continue reading
-      if( this.push( buffer ) ) {
-        this._isReading = false
-        // If we've hit the end of the range,
-        // proceed to the next range
-        if( this.blockInRange >= this.rangeBlockCount ) {
-          this._read()
-        } else {
-          this._readBlock()
-        }
-      }
+      this.push( buffer )
 
     })
 
   },
 
+  /**
+   * Initiate a new read, advancing the range if necessary,
+   * and verifying checksums, if enabled
+   * @see https://nodejs.org/api/stream.html#stream_implementing_a_readable_stream
+   * @private
+   */
   _read() {
 
-    if( this._isReading ) return;
-
     if( this.fd == null ) {
-      return this._open( this._read )
+      return void this.once( 'open', () => { this._read() })
     }
 
-    if( this.rangesRead === this.blockMap.ranges.length ) {
-      return this._close()
+    if( this.ranges.length === 0 ) {
+      return void this.push( null )
     }
 
-    // If we're either reading the first block, or have completed a block range;
-    // verify, then switch to next range & continue block reads
-    if( this.currentRange == null || this.blockInRange >= this.rangeBlockCount ) {
+    if( this.currentRange == null || this.currentRange.offset === this.currentRange.length ) {
       this._verifyRange()
-      this.currentRange = this.blockMap.ranges[ this.rangesRead++ ]
-      this.blockInRange = 0
-      this._readBlock()
+      this.currentRange = this.ranges.shift()
+      this.rangesRead++
+      debug( 'read:range %O', this.currentRange )
     }
 
+    this._readBlock()
+
   },
 
-  destroy() {
+  /**
+   * Open a handle to the image file in question
+   * @private
+   */
+  open() {
 
-    if( !this.options.autoClose ) {
-      stream.Readable.prototype.destroy.call( this )
+    if( this.fd != null ) {
       return
     }
 
+    fs.open( this.path, this.flags, this.mode, ( error, fd ) => {
+      if( error ) {
+        if( this.autoClose ) {
+          this.destroy()
+        }
+        this.emit( 'error', error )
+      } else {
+        this.fd = fd
+        this.emit( 'open', this.fd )
+      }
+    })
+
+  },
+
+  /**
+   * Close the stream, and the underlying file handle
+   * @param {Function} [callback] - callback(error)
+   */
+  close( callback ) {
+
+    if( callback ) {
+      this.once( 'close', callback )
+    }
+
+    if( this.closed || this.fd == null ) {
+      if( this.fd == null ) {
+        this.once( 'open', () => {
+          this.close()
+        })
+      } else {
+        process.nextTick(() => {
+          this.emit( 'close' )
```

Comment:
Not sure if it is intended but calling `close()` twice could lead to 2 `close` events fired: one here and one in `fs.close()` callback.